lib/concurrent/atomic/thread_local_var.rb

```diff
@@ -33,169 +33,131 @@ module Concurrent
   #
   #   @see https://docs.oracle.com/javase/7/docs/api/java/lang/ThreadLocal.html Java ThreadLocal
   #
-  # @!visibility private
-  class AbstractThreadLocalVar
+  class ThreadLocalVar
+
+    # Each thread has a (lazily initialized) array of thread-local variable values
+    # Each time a new thread-local var is created, we allocate an "index" for it
+    # For example, if the allocated index is 1, that means slot #1 in EVERY
+    #   thread's thread-local array will be used for the value of that TLV
+    #
+    # The good thing about using a per-THREAD structure to hold values, rather
+    #   than a per-TLV structure, is that no synchronization is needed when
+    #    reading and writing those values (since the structure is only ever
+    #    accessed by a single thread)
+    #
+    # Of course, when a TLV is GC'd, 1) we need to recover its index for use
+    #   by other new TLVs (otherwise the thread-local arrays could get bigger
+    #   and bigger with time), and 2) we need to null out all the references
+    #   held in the now-unused slots (both to avoid blocking GC of those objects,
+    #   and also to prevent "stale" values from being passed on to a new TLV
+    #   when the index is reused)
+    # Because we need to null out freed slots, we need to keep references to
+    #   ALL the thread-local arrays -- ARRAYS is for that
+    # But when a Thread is GC'd, we need to drop the reference to its thread-local
+    #   array, so we don't leak memory
 
     # @!visibility private
     NIL_SENTINEL = Object.new
-    private_constant :NIL_SENTINEL
+    FREE = []
+    LOCK = Mutex.new
+    ARRAYS = {} # used as a hash set
```

Comment:
NO_VALUE (NIL_SENTINEL) should never be used externally to this class. Internally, it indicates "this variable has been explicitly set to `nil`". `nil` indicates "this variable has never been explicitly set yet, so its value is still the default". When you call `#value` or `#value=`, it converts accordingly, so only `nil` is seen by external callers.

I will change NIL_SENTINEL to NO_VALUE, and make the constants private.